engine/runtime/src/main/java/org/enso/interpreter/node/expression/builtin/bool/OrNode.java

```diff
@@ -1,14 +1,41 @@
 package org.enso.interpreter.node.expression.builtin.bool;
 
 import com.oracle.truffle.api.nodes.Node;
+import com.oracle.truffle.api.profiles.ConditionProfile;
 import org.enso.interpreter.dsl.BuiltinMethod;
+import org.enso.interpreter.dsl.MonadicState;
+import org.enso.interpreter.dsl.Suspend;
+import org.enso.interpreter.node.BaseNode;
+import org.enso.interpreter.node.callable.thunk.ThunkExecutorNode;
+import org.enso.interpreter.runtime.Context;
+import org.enso.interpreter.runtime.error.DataflowError;
+import org.enso.interpreter.runtime.error.PanicException;
+import org.enso.interpreter.runtime.state.Stateful;
 
 @BuiltinMethod(
     type = "Boolean",
     name = "||",
     description = "Computes the logical disjunction of two booleans")
 public class OrNode extends Node {
-  boolean execute(boolean _this, boolean that) {
-    return _this || that;
+  private @Child ThunkExecutorNode rhsThunkExecutorNode = ThunkExecutorNode.build();
+  private final ConditionProfile condProfile = ConditionProfile.createCountingProfile();
+
+  Stateful execute(@MonadicState Object state, boolean _this, @Suspend Object that) {
+    if (condProfile.profile(_this)) {
```

Comment:
What is the intention behind [profiling using counting profile](https://www.graalvm.org/truffle/javadoc/com/oracle/truffle/api/profiles/ConditionProfile.html#createCountingProfile--)? Should `||` and `&&` on their own act as an `if` branching point?

Reply:
That was the idea but probably an unnecessary one, especially considering your second comment